homeassistant/components/statistics/sensor.py

```diff
@@ -474,37 +472,29 @@ def _scheduled_update(now: datetime) -> None:
     def _fetch_states_from_database(self) -> list[State]:
         """Fetch the states from the database."""
         _LOGGER.debug("%s: initializing values from the database", self.entity_id)
-        states = []
-
-        with session_scope(hass=self.hass) as session:
-            query = session.query(States, StateAttributes).filter(
-                States.entity_id == self._source_entity_id.lower()
-            )
-
-            if self._samples_max_age is not None:
-                records_older_then = dt_util.utcnow() - self._samples_max_age
-                _LOGGER.debug(
-                    "%s: retrieve records not older then %s",
-                    self.entity_id,
-                    records_older_then,
-                )
-                query = query.filter(States.last_updated >= records_older_then)
-            else:
-                _LOGGER.debug("%s: retrieving all records", self.entity_id)
-
-            query = query.outerjoin(
-                StateAttributes, States.attributes_id == StateAttributes.attributes_id
+        lower_entity_id = self._source_entity_id.lower()
+        if self._samples_max_age is not None:
+            start_date = (
+                dt_util.utcnow() - self._samples_max_age - timedelta(microseconds=1)
             )
-            query = query.order_by(States.last_updated.desc()).limit(
-                self._samples_max_buffer_size
+            _LOGGER.debug(
+                "%s: retrieve records not older then %s",
+                self.entity_id,
+                start_date,
             )
-            if results := execute(query, to_native=False, validate_entity_ids=False):
-                for state, attributes in results:
-                    native = state.to_native()
-                    if not native.attributes:
-                        native.attributes = attributes.to_native()
-                    states.append(native)
-        return states
+        else:
+            start_date = datetime.fromtimestamp(0, tz=dt_util.UTC)
+            _LOGGER.debug("%s: retrieving all records", self.entity_id)
+        entity_states = history.state_changes_during_period(
+            self.hass,
+            start_date,
+            entity_id=lower_entity_id,
+            descending=True,
+            limit=self._samples_max_buffer_size,
+            include_start_time_state=False,
+        )
+        # Need to cast since minimal responses is not passed in
```

Comment:
What does this comment mean?
Also, why is the return value of `state_changes_during_period` as flexible as `MutableMapping[str, Iterable[LazyState | State | dict[str, Any]]]` yet we can just cast it to `list[State]`?

Reply:
If minimal_responses is True it will return the intermediate states as dicts.  In the future we could split it into two functions to clean up the typing

Reply:
What is `minimal_responses`, there's no such parameter in `history.state_changes_during_period`?

Reply:
We could cast it in `state_changes_during_period` as its inheriting the typing from:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/recorder/history.py#L507

Reply:
Let me know if you prefer that or keeping it as-is and I'll make the change

Reply:
Yes, let's correct the typing of `state_changes_during_period` (and  `get_last_state_changes`) since we know they will never include list elements.

Reply:
https://github.com/home-assistant/core/pull/68917/files

Reply:
https://github.com/home-assistant/core/pull/68917/files

Reply:
Looks like it still has to be cast for now since we call reverse on it.  Once we do the refactor in the other PR to split it we can take it out